File: Source/GobangGameLib/GameJudge/BasicJudge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using GobangGameLib.GameBoard;
using GobangGameLib.GameBoard.PositionManagement;

namespace GobangGameLib.GameJudge
{
    public class BasicJudge : IJudge
    {
        private readonly BoardProperties _context;
        private readonly PositionManager _positions;

        public BasicJudge(BoardProperties context, PositionManager positions)
        {
            _context = context;
            _positions = positions;
        }

        public PieceType GetWinner(IBoard board)
        {
            var winner = _positions.Lines.Select(line => GetWinnerInOneLine(board, line))
                .Where(winnerPiece => !winnerPiece.Equals(PieceType.Empty))
                .FirstOrDefault();

            return winner;
        }

        private PieceType GetWinnerInOneLine(IBoard board, IPositions line)
        {
            var count = 0;
            var pre = PieceType.Empty;

            foreach (var position in line.Positions)
            {
                var piece = board.Get(position);
                if (!pre.Equals(PieceType.Empty) && pre.Equals(piece))
                {
                    count++;
                    if (count >= _context.NumOfPiecesToWin)
                    {
                        return piece;
                    }
                }
                else if (!pre.Equals(piece))
                {
                    count = 1;
                }
                pre = piece;
            }

            return PieceType.Empty;
        }
    }
}

File: Java/GobangCore/src/com/tiger/gobangcore/models/BasicJudge.java
package com.tiger.gobangcore.models;

import com.tiger.gobangcore.interfaces.Board;
import com.tiger.gobangcore.interfaces.Judge;

public class BasicJudge implements Judge {

    @Override
    public PieceType GetWinner(Board board) {
        PieceType winner = this.checkRow(board);
        if (winner != PieceType.EMPTY) {
            return winner;
        }

        winner = this.checkCol(board);
        if (winner != PieceType.EMPTY) {
            return winner;
        }

        return PieceType.EMPTY;
    }

    private PieceType checkRow(Board board) {
        int connectedCount;
        PieceType winner;
        for (int r = 0; r < board.getBoardContext().rowSize; r++) {
            connectedCount = 0;
            winner = PieceType.EMPTY;
            for (int c = 0; c < board.getBoardContext().colSize; c++) {
                Position p = new Position(r, c);
                PieceType currentPiece = board.get(p);
                if (currentPiece == PieceType.EMPTY) {
                    winner = PieceType.EMPTY;
                    connectedCount = 0;
                } else {
                    if (currentPiece != winner) {
                        connectedCount = 0;
                    }

                    connectedCount++;
                    winner = currentPiece;

                    if (connectedCount == board.getBoardContext().numOfPiecesToWin) {
                        return winner;
                    }
                }
            }
        }

        return PieceType.EMPTY;
    }

    private PieceType checkCol(Board board) {
        int connectedCount;
        PieceType winner;
        for (int c = 0; c < board.getBoardContext().colSize; c++) {
            connectedCount = 0;
            winner = PieceType.EMPTY;
            for (int r = 0; r < board.getBoardContext().rowSize; r++) {
                Position p = new Position(r, c);
                PieceType currentPiece = board.get(p);
                if (currentPiece == PieceType.EMPTY) {

                    winner = PieceType.EMPTY;
                    connectedCount = 0;
                } else {
                    if (currentPiece != winner) {
                        connectedCount = 0;
                    }

                    connectedCount++;
                    winner = currentPiece;

                    if (connectedCount == board.getBoardContext().numOfPiecesToWin) {
                        return winner;
                    }
                }
            }
        }

        return PieceType.EMPTY;
    }
}
